src/mds/MDSPerfMetricTypes.h

```diff
@@ -0,0 +1,189 @@
+    os << "{latency=" << metric.lat << "}";
+    return os;
+  }
+};
+
+struct WriteLatencyMetric {
+  utime_t lat;
+
+  DENC(WriteLatencyMetric, v, p) {
+    DENC_START(1, 1, p);
+    denc(v.lat, p);
+    DENC_FINISH(p);
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const WriteLatencyMetric &metric) {
+    os << "{latency=" << metric.lat << "}";
+    return os;
+  }
+};
+
+struct MetadataLatencyMetric {
+  utime_t lat;
+
+  DENC(MetadataLatencyMetric, v, p) {
+    DENC_START(1, 1, p);
+    denc(v.lat, p);
+    DENC_FINISH(p);
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const MetadataLatencyMetric &metric) {
+    os << "{latency=" << metric.lat << "}";
+    return os;
+  }
+};
+
+WRITE_CLASS_DENC(CapHitMetric)
+WRITE_CLASS_DENC(ReadLatencyMetric)
+WRITE_CLASS_DENC(WriteLatencyMetric)
+WRITE_CLASS_DENC(MetadataLatencyMetric)
+
+// metrics that are forwarded to the MDS by client(s).
+struct Metrics {
+  // metrics
+  CapHitMetric cap_hit_metric;
+  ReadLatencyMetric read_latency_metric;
+  WriteLatencyMetric write_latency_metric;
+  MetadataLatencyMetric metadata_latency_metric;
+
+  // metric update type
+  UpdateType update_type = UpdateType::UPDATE_TYPE_REFRESH;
+
+  void encode(bufferlist &bl) const {
+    using ceph::encode;
+
+    ENCODE_START(1, 1, bl);
+    encode(static_cast<uint32_t>(update_type), bl);
+    encode(cap_hit_metric, bl);
+    encode(read_latency_metric, bl);
+    encode(write_latency_metric, bl);
+    encode(metadata_latency_metric, bl);
+    ENCODE_FINISH(bl);
+  }
+
+  void decode(bufferlist::const_iterator &iter) {
+    using ceph::decode;
+
+    DECODE_START(1, iter);
+    uint32_t type;
+    decode(type, iter);
+    update_type = static_cast<UpdateType>(type);
+    decode(cap_hit_metric, iter);
+    decode(read_latency_metric, iter);
+    decode(write_latency_metric, iter);
+    decode(metadata_latency_metric, iter);
+    DECODE_FINISH(iter);
+  }
+
+  void dump(Formatter *f) const {
+    f->dump_int("update_type", static_cast<uint32_t>(update_type));
+  }
+
+  friend std::ostream& operator<<(std::ostream& os, const Metrics& metrics) {
+    os << "[update_type=" << metrics.update_type << ", metrics={"
+       << "cap_hit_metric=" << metrics.cap_hit_metric
+       << ", read_latency=" << metrics.read_latency_metric
+       << ", write_latency=" << metrics.write_latency_metric
+       << ", metadata_latency=" << metrics.metadata_latency_metric
+       << "}]";
+    return os;
+  }
+};
+
+// Initial Sequence Number (ISN) of the first ping message sent
+// by rank 0 to other active ranks (incuding itself).
+#define MDS_PINGER_ISN 1
+
+struct MetricsMessage {
+  version_t seq = 0;
+  mds_rank_t rank = MDS_RANK_NONE;
+  std::map<entity_inst_t, Metrics> client_metrics_map;
+
+  MetricsMessage() {
+  }
+  MetricsMessage(version_t seq, mds_rank_t rank)
+    : seq(seq), rank(rank) {
+  }
+
+  void encode(bufferlist &bl, uint64_t features) const {
```

Comment:
nit, could use `DENC` macro instead. like
```c++
DENC(MetricsMessage, v, p) {
  DENC_START(1, 1, p);
  denc(v.seq, p);
  /// ...
  DENC_FINISH(p);
}
```

Reply:
`client_metrics_map` field in this class uses the old-style encoding. Is it ok for this class to use new-style encoding? I tried this in the past but couldn't get it working.

Reply:
i don't see a reason why we cannot start using new-style encoding here. 

> but couldn't get it working.

could you elaborate a little bit?

Reply:
> i don't see a reason why we cannot start using new-style encoding here.
> 
> > but couldn't get it working.
> 
> could you elaborate a little bit?

well, it was a while back -- do not remember the specifics, but somewhat relating to `denc` macro unable to encode/decode member variable using old-style encoders.

I'll try again.

Reply:
@tchaikov -- need a bit of help in converting to new-style encoding defs:

For this class, I changed it as per:

```
+  DENC_FEATURED(metrics_message_t, v, p, f) {
+    DENC_START(1, 1, p);
+    denc(v.seq, p);
+    denc(v.rank, p);
+    denc(v.client_metrics_map, p, f);
+    DENC_FINISH(p);
+  }

+WRITE_CLASS_DENC_FEATURED(metrics_message_t)
```
(_FEATURED version is required since `entity_inst_t` need to be encoded).

With this change I get the following compilation err:

```
/work/ceph/src/mds/MDSPerfMetricTypes.h: In instantiation of ‘std::enable_if_t<(is_same_v<T, metrics_message_t> || is_same_v<T, const metrics_message_t>)> _denc_friend(T&, P&, uint64_t) [with T = const metrics_message_t; P = long unsigned int; std::enable_if_t<(is_same_v<T, metrics_message_t> || is_same_v<T, const metrics_message_t>)> = void; uint64_t = long unsigned int]’:
/work/ceph/src/mds/MDSPerfMetricTypes.h:132:3:   required from here
/work/ceph/src/mds/MDSPerfMetricTypes.h:136:9: error: no matching function for call to ‘denc(const std::map<entity_inst_t, Metrics>&, long unsigned int&, uint64_t&)’
     denc(v.client_metrics_map, p, f);
     ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I tried looking at include/denc.h but couldn't get a deep understanding of the macros.  So, for testing, I change the std::map to be `<uint32_t, Metrics>` and that compiles fine. Seems to me that some additional handling needs to be done for `entity_inst_t`?

Reply:
@batrick ^^

Reply:
Hmm, you're in the weeds aren't you. I'm not sure. Let's wait and see if @tchaikov knows off-hand what's wrong.

Reply:
@vshankar sorry, i was wrong -- the new-style denc requires the member variables to be encoded are also properly annotated with `denc_traits<>` definitions. and to switch `entity_addr_t` to denc machinery is not trivial. so i am resolving this conversation.